[WIP] Add on screen kick message expr

nopeless opened the pull request:
### Description

Implement kick message getting and setting

---
**Target Minecraft Versions:** Not known yet

**Related Issues:**

https://discord.com/channels/135877399391764480/1332620961328926790

Aresiel commented:
OMG I literally had this issue today and this PR was made like the second I found it. What are the odds.

nopeless commented:
Annotations need proper skript version needs to be updated, hence the WIP

Burbulinis commented:
This PR should target dev/feature, not master

Fusezion commented:
This should already be possible via ExprMessage https://github.com/SkriptLang/Skript/blob/2bccb3232f8da54cece796ba277d1355f8a83d0b/src/main/java/ch/njol/skript/expressions/ExprMessage.java#L83-L100

sovdeeth commented:
> Annotations need proper skript version needs to be updated, hence the WIP

Any additions should use `INSERT VERSION` as a placeholder. It's automatically replaced during release

nopeless commented:
> This should already be possible via ExprMessage
> 
> https://github.com/SkriptLang/Skript/blob/2bccb3232f8da54cece796ba277d1355f8a83d0b/src/main/java/ch/njol/skript/expressions/ExprMessage.java#L83-L100

@Fusezion 
This turns out to set the message in the chat not the kick message itself, hence the change

Fusezion commented:
> 
> @Fusezion This turns out to set the message in the chat not the kick message itself, hence the change

Hmm, alright.

nopeless commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,83 @@
+@RequiredPlugins("xxx Paper 1.16.5+")
+@Since("xxx 2.8.0")
+@Events("Kick")
+public class ExprKickMessage extends SimpleExpression<String> {
```
working on the method order;

I don't think it should be part of `ExprMessage`, its not really a message that appears in chat. The reason why I named it `ExprKickMessage` is because kick reason was already taken

sovdeeth replied:
there's no ExprKickReason, though?

nopeless replied:
... good point, but then the class name would be `ExprKickReason` while the actual pattern is `kick message` and I thought that was worse

sovdeeth replied:
Why would the pattern stay as kick message? shouldn't it be kick reason too?

nopeless commented:
@sovdeeth If you have a class name recommendation let me know because its very ambiguous overall

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+public class ExprKickMessage extends SimpleExpression<String> {
+
+	static {
+		if (Skript.classExists("org.bukkit.event.player.PlayerKickEvent"))
```
As of Skript 2.10, we only support 1.19.4+
So this check is not needed

Efnilite replied:
```suggestion
```
this class has existed since pre-1.13

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	protected String @Nullable [] get(Event event) {
+		if (event instanceof PlayerKickEvent playerKickEvent) {
+			return new String[] { MiniMessage.miniMessage().serialize(playerKickEvent.reason()) };
```
We should be checking if this method exists.
Paper deprecated `#getReason` in favor of their own `#reason` 
So we need to check if `#reason` exists, if not, then we use `#getReason`

Fusezion replied:
With the fact `#reason` is text compnent, I do not believe we should be using it, as far as I'm aware his code here would fail on spigot servers unless they implement/shade the mini message api

Efnilite replied:
no, use getReason/setReason as no other thing in skript uses adventure rn

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+		assert delta.length == 1;
+		if (e instanceof PlayerKickEvent event && delta[0] instanceof String text) {
+			// avoid deprecated api
+			event.reason(MiniMessage.miniMessage().deserialize(text));
```
Same as above, if `#reason` does not exist, then we use `#setReason`

Efnilite replied:
```suggestion
			event.setReason(text);
```

TheAbsolutionism reviewed (changes requested):
I know Sovde already mentioned about the coding conventions, so I won't do anything related to that.

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		return getParser().isCurrentEvent(PlayerKickEvent.class);
```
```suggestion
		return true;
```
use `EventRestrictedSyntax` :)

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	protected String @Nullable [] get(Event event) {
+		if (event instanceof PlayerKickEvent playerKickEvent) {
+			return new String[] { MiniMessage.miniMessage().serialize(playerKickEvent.reason()) };
```
use the deprecated getReason method here, as minimessage components are not used anywhere else in skript

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(final Changer.ChangeMode mode) {
```
```suggestion
	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
```

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+	@Override
```
```suggestion
	}

	@Override
```

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "the kick reason";
+	}
+}
```
```suggestion
	}

}
```

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	@Override
+	public void change(Event e, final Object[] delta, Changer.ChangeMode mode) {
```
```suggestion
	public void change(Event event, Object[] delta, ChangeMode mode) {
```

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+import org.jetbrains.annotations.Nullable;
+
+@Name("Kick Message")
+@Description(
+	"The <a href='classes.html#kickreason'>kick message</a> as to why a player was kicked " +
+	"(does not appear in chat but the kicked player's screen)<a href='events.html#kick'>kick</a> event.")
```
```suggestion
@Description("The kick message that is displayed on-screen when a player is kicked.")
```

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+		"\tkick message is 'Invalid hotbar selection (Hacking?)'",
+		"\tcancel event"
+})
+@RequiredPlugins("xxx Paper 1.16.5+")
```
```suggestion
```

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+public class ExprKickMessage extends SimpleExpression<String> {
+
+	static {
+		Skript.registerExpression(ExprKickMessage.class, String.class, ExpressionType.SIMPLE, "(disconnect|kick) message");
```
```suggestion
		Skript.registerExpression(ExprKickMessage.class, String.class, ExpressionType.SIMPLE, "kick (message|reason)");
```
There's a difference between disconnecting and getting kicked, so I don't think `disconnect` should be in the pattern
Also, since the methods that are being utilized is **reason** should also allow `reason` in the pattern

Fusezion replied:
`kick message` would conflict with the existing `kick message` expression under ExprMessage

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+	}
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
```
```suggestion
	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```

Fusezion replied:
\*imports required

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] @Nullable acceptChange(final ChangeMode mode) {
```
```suggestion
	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
```

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+	}
+
+	@Override
+	public void change(Event e, final Object[] delta, ChangeMode mode) {
```
```suggestion
	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
```
Make sure to change the `e`, bad practice to have 1 letter variables, imo atleast

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+	public void change(Event e, final Object[] delta, ChangeMode mode) {
+		assert delta != null;
+		assert delta.length == 1;
+		if (e instanceof PlayerKickEvent event && delta[0] instanceof String text) {
+			event.setReason(text);
+		}
```
```suggestion
		if (event instanceof PlayerKickEvent kickEvent && delta[0] instanceof String text) {
			kickEvent.setReason(text);
		}
```

TheAbsolutionism commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,74 @@
+})
+@Since("INSERT VERSION")
+@Events("Kick")
+public class ExprKickMessage extends SimpleExpression<String> {
```
Aas per Efy's requested change, you need to implement `EventRestrictedSyntax`
Without it, then this element would not throw a parse error when used outside of an `on kick` event

ShaneBeee commented:
In my opinion, rather than creating another expression, this should probably just be merged in with:
https://docs.skriptlang.org/expressions.html?search=#ExprMessage

nopeless commented:
@ShaneBeee ExprMessage is what gets sent in chat, no? This is for what gets displayed on the player's screen

ShaneBeee commented:
> @ShaneBeee ExprMessage is what gets sent in chat, no? This is for what gets displayed on the player's screen

ohh ok gotcha. Anyways, wont your expression collide with `kick message` from ExprMessage?

nopeless commented:
@ShaneBeee It does, currently, I'm not sure how to resolve it. My friends and people here has obviously said that all of this is ambiguous and I am leaving it up to maintainers to decide on the pattern and name

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+public class ExprKickMessage extends SimpleExpression<String> implements EventRestrictedSyntax {
+
+	static {
+		Skript.registerExpression(ExprKickMessage.class, String.class, ExpressionType.SIMPLE, "kick (message|reason)");
```
```suggestion
		Skript.registerExpression(ExprKickMessage.class, String.class, ExpressionType.SIMPLE, "(display[ed]|on-screen) kick (message|reason)");
``` there should be a way to differentiate the chat kick message from this one

Efnilite commented on src/main/java/ch/njol/skript/expressions/ExprKickMessage.java:
```diff
@@ -0,0 +1,78 @@
+import ch.njol.skript.doc.*;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Kick Message")
```
```suggestion
@Name("On-screen Kick Message")
```

nopeless commented:
@Efnilite I'm guessing that the expression pattern is final? (I don't know if you have discussed with others)

Fusezion commented:
> @Efnilite I'm guessing that the expression pattern is final? (I don't know if you have discussed with others)

Just providing my two cents, I would prefer going with Efy's changes here as that not only fixes conflictions but is descriptive enough that players can tell the difference between this and it's counter part

Efnilite commented:
> @Efnilite I'm guessing that the expression pattern is final? (I don't know if you have discussed with others)

if you have better suggestions feel free to use them, we just need to make sure its different

nopeless commented:
I'll leave this open for a bit just in case someone wants to pitch in. personally I don't care what the pattern is as long as its usable

sovdeeth commented:
you can re-request me once requested changes are addressed

nopeless commented:
few notes: I renamed file to `ExprOnScreenKickMessage` so it is not ambiguous at all. Updated example and also set the pattern to just a single literal.

erenkarakal commented on src/main/java/ch/njol/skript/expressions/ExprOnScreenKickMessage.java:
```diff
@@ -0,0 +1,79 @@
+@Description("The kick message that is displayed on-screen when a player is kicked.")
+@Examples({
+	"on kick:",
+		"\ton-screen kick message is 'Invalid hotbar selection (Hacking?)'",
```
single quotes don't exist in Skript

erenkarakal commented on src/main/java/ch/njol/skript/expressions/ExprOnScreenKickMessage.java:
```diff
@@ -0,0 +1,79 @@
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "the kick reason";
```
this should be `on-screen kick message`

sovdeeth commented on src/main/java/ch/njol/skript/expressions/ExprOnScreenKickMessage.java:
```diff
@@ -0,0 +1,79 @@
+public class ExprOnScreenKickMessage extends SimpleExpression<String> implements EventRestrictedSyntax {
+
+	static {
+		Skript.registerExpression(ExprOnScreenKickMessage.class, String.class, ExpressionType.SIMPLE, "on-screen kick message");
```
```suggestion
		Skript.registerExpression(ExprOnScreenKickMessage.class, String.class, ExpressionType.SIMPLE, "[the] on-screen kick message");
```

sovdeeth commented on src/main/java/ch/njol/skript/expressions/ExprOnScreenKickMessage.java:
```diff
@@ -0,0 +1,79 @@
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "the on-screen kick reason";
```
```suggestion
		return "the on-screen kick message";
```
